### PR TITLE
Fix type hints for render and on

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

I noticed two things that were off about the type hints for on and render
- We were using `Any` for the triggers of `gr.on`. Which was too broad.
- In `render`, the triggers were not typed correctly so passing a valid trigger, e.g. `slider.change` would be flagged in vs code. 
- In `render`, the type hint did not allow for a single component to be passed as input, but it is allowed.

**Main**
Screenshot 2024-05-31 at 2 09 04 PM
Screenshot 2024-05-31 at 2 03 46 PM

**PR**
Invalid value for `triggers` of `gr.on` is flagged
image

Passing an event listener method for `render` is not flagged
image

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/renderable.py
Comment (abidlabs) [inline gradio/renderable.py:83]: The only issue with putting `Sequence` here is that it'll affect the docs for gr.render(): https://www.gradio.app/docs/gradio/render

Comment (abidlabs) [reply gradio/renderable.py:83]: And I think many Python users don't know what `Sequence` is

Comment (abidlabs) [reply gradio/renderable.py:83]: Can we add a step in the docs generation to convert `Sequence` container types to `list` @aliabd?

**Review by aliabid94 — APPROVED** [review]
LGTM! I was trying to see if we could actually use EventListenerCallable to typehint EventListener.listener, so that if EventListenerCallable ever diverges from the actual signature of a listener, our typecheker catches it. Cou;dn't figure it out though

File: gradio/renderable.py
Comment (freddyaboulton) [reply gradio/renderable.py:83]: I think I can edit `01_render.svx` to do this but maybe we do this once users complain? I would prefer to not introduce a difference between docs and code unless it's absolutely necessary imo.

Comment (abidlabs) [reply gradio/renderable.py:83]: Ok I think this would be applicable in many other parts of the codebase as well (e.g. anywhere we currently type hint list[Component] should really be Sequence[Component]) but doing it later is fine

**Review by abidlabs — APPROVED** [review]
LGTM

File: gradio/renderable.py
Comment (freddyaboulton) [reply gradio/renderable.py:83]: Ok I think we can edit the ParamTable svelte component in that case but will hold off for now.

**Comment by freddyaboulton** [comment]
Thanks for the reviews everyone!

**Pull request merged**
